base/reinterpretarray.jl

```diff
@@ -0,0 +1,67 @@
+"""
+Gives a reinterpreted view (of element type T) of the underlying array (of element type S).
+If the size of `T` differs from the size of `S`, the array will be compressed/expanded in
+the first dimension.
+"""
+struct ReinterpretArray{T,S,N,A<:AbstractArray{S, N}} <: AbstractArray{T, N}
+    parent::A
+    Base.reinterpret(::Type{T}, a::A) where {T,S,N,A<:AbstractArray{S, N}} = new{T, S, N, A}(a)
+end
+
+Base.eltype(a::ReinterpretArray{T}) where {T} = T
+function Base.size(a::ReinterpretArray{T,S}) where {T,S}
+    psize = size(a.parent)
+    if sizeof(T) > sizeof(S)
+        size1 = div(psize[1], div(sizeof(T), sizeof(S)))
```

Comment:
Presumably this first `div` can't be const-ified by inference/LLVM. Since `size` is used in bounds-checks, and since `div` is achingly slow, I wonder if we need to cache this value in the `struct`.

Reply:
`div(T, S)` is constant folded. The other div is generally folded to shifts by LLVM.